[그리디] 염지환 로또 미션 1,2 단계 제출합니다

JihwanYeom opened the pull request:
안녕하세요 리뷰어님! 이번에 로또 미션 리뷰 받게 된 염지환 이라고 합니다! 부족한 부분이 많아
지적을 받을수록 많이 배워가고 있는 것 같습니다. 조금이라도 잘못된 것 같은 코드가 있으면 얘기해 주시면
감사하겠습니다!

이번 미션에서 가장 어려웠던 건 모든 원시값을 포장하라는 조건과 일급 컬렉션을 사용하라는 조건이었던 것 같습니다
제가 아직 원시값 포장에 익숙하지 않아서 그런지 포장된 객체를 사용하면서 객체지향에 맞지 않는 코드를 짜고 있다는
느낌이 많이 들었습니다.

예를 들면 모든 값을 포장해두면 getter를 자연스럽게 쓰게 되고 이로 인해 해당 객체를 활용하는 코드가 많이 길어진다는
느낌을 받았습니다. 제가 객체의 역할을 잘못 부여해서 생기는 일일까요?

그리고 포장된 객체들의 리스트를 다루는 것도 고민이 많이 되었습니다 포장된 객체들의 리스트를 가공하거나
출력하기 위해 stream을 많이 사용했는데 이것이 정말 좋은 코드인지, 너무 복잡하게 짜는 코드는 아닌지
고민했던 것 같습니다.

그리고 Rank에 대한 가격을 상수로 선언해 두었는데 이렇게 어떤 수에 따른 결과값을 상수로 선언하는 것이 좋은
방법은 아닌 것 같습니다.

원래는 테스트 코드 까지 작성하려고 했으나 로직을 고민하는데 시간을 많이 사용하여 아직 테스트코드를 작성하지
못했습니다 빠른 시일 내로 전까지 단위 테스트 코드까지 완성해 보도록 하겠습니다.
README 역시 아직 많은 내용을 적지 못하였습니다. 빠른 시일 내로 README도 작성하겠습니다...

마지막으로 이번 미션 잘 부탁드립니다! 감사합니다!

JihwanYeom committed:
feat : 로또 숫자 생성 로직 구현

LottoNumberGenerator 인터페이스 생성

LottoNumberGenerator 의 구현체
RandomLottoNumberGenerator 생성

1에서 45 사이의 숫자 중 중복없이
숫자 6개를 뽑아 리스트로 만들어
반환.

JihwanYeom committed:
feat : 로또 클래스 구현

생성자에서 로또 번호를 생성하여
로또 번호를 저장하고 있는 로또
클래스 구현

JihwanYeom committed:
feat : View 의 입출력 메서드 구현

금액을 입력받는 readPrice 와
로또 번호들을 출력하는 printResult
메서드 구현

JihwanYeom committed:
feat : Controller 및 실행파일 구현

View 를 통해 금액을 입력받아
로또의 개수만큼 로또를 생성하고
생성된 로또들의 번호를 출력하는
컨트롤러 구현

실행을 위한 main 메서드 작성

JihwanYeom committed:
refactor : 원시값 포장, 일급 컬렉션, 정적 팩토리 메서드 적용

LottoNumber, Price 클래스를 만들어 원시값 포장

Lottos 일급 컬렉션 구현

각 객체에 정적 팩토리 메서드 구현

JihwanYeom committed:
feat : 2단계 기능 구현

사용자로부터 당첨 번호를 입력받는 기능 구현

당첨 번호를 통해 각 로또의 당첨 번호 개수 리스트를 반환하는 기능 구현

당첨 번호 개수 리스트를 통해 총 상금을 구하는 기능 구현

결과를 출력하는 기능 구현

JihwanYeom committed:
docs : README.md 개요 작성

프로젝트 개요 작성

JihwanYeom committed:
style : stream 가독성 향상

지역변수를 활용한 stream 가독성 향상

JihwanYeom committed:
style : 코드 서식 점검

JihwanYeom committed:
chore : 푸시 전 경고 수정

JihwanYeom committed:
chore : 푸시 전 경고 수정2

JihwanYeom committed:
test : 단위 테스트 구현

중요한 기능을 가진 로직에들 대해 단위 테스트 구현.

70825 commented on README.md:
```diff
@@ -0,0 +1,13 @@
+# 💰 자바 로또
```
> 원래는 테스트 코드 까지 작성하려고 했으나 로직을 고민하는데 시간을 많이 사용하여 아직 테스트코드를 작성하지
못했습니다 빠른 시일 내로 전까지 단위 테스트 코드까지 완성해 보도록 하겠습니다.

로직을 고민하는데 시간을 많이 사용한 근본적인 원인이 무엇이라고 생각하시나여?? (정말 큰 시간을 차지하는 곳)
요거는 말씀해주시면 같이 고민해볼만한 것 같아요~
(ex. 이미 만든 로직도 처음부터 다시 고민하는걸 반복하게 되는지, 아니면 큰 그림을 잡으려고 하는데 시간이 오래 걸리는지, ...)

JihwanYeom replied:
제가 로직을 고민하는 데 시간을 많이 썼던 이유는 너무 생각을 열고 코드를 작성하지 못해서라고 생각합니다. 예를 들면 저는 이번 미션에서 다음과 같이 생각했습니다.

Lotto 랑 Lottos 객체에 LottoRank와 LottoRanks 객체를 인스턴스 변수로 추가하여 작업을 하면 편하겠지만 인스턴스 변수를 추가하는 것은 지양해야겠다고 판단하여 Rank를 변수로 만들지 않고, 단순히 Lotto와 Lottos 객체에 Rank와 관련된
메서드만을 만들어 코드를 작성하였습니다. 이 과정에서 메서드로만 객체 간의 정보를 주고받다 보니 스스로 로직을 떠올리기 힘든 부분이 찾아왔습니다.

그러다 얼마 전 미션을 진행한 다른 동아리 팀원 분과 얘기하면서 그 분은 Rank를 구하기 위해 아예 로또 순위를 결정하는 역할을 가진 클래스를 하나 더 만드셨다는 얘기를 듣고 제가 생각이 너무 닫혀있었던 건 아닌지 생각했습니다.

코드를 작성하면서 유연한 사고를 하는 건 당연한 것인데 아직 저는 하나를 확실히 하려다가 다른 것을 놓치는 단계에 머물러 있는 것 같습니다. 특히 저도 모르게 변수나 클래스를 최대한 적게 만들어야 한다는 강박이 있는 것 같습니다. 학교 수업에서 코드를 작성하면서 생긴 좋지 않은 습관 같기도 합니다...

꾸준한 연습이 답이라는 걸 저도 알고 있지만. 코드를 작성하면서 유연하게 사고하는 능력을 키우기 위해 제가 더 신경써야 할 부분이 있을까요??

70825 replied:
음.. 그대로만 해도 되지 않을까요..?? 저는 잘하고 있어보여요

> 코드를 작성하면서 유연하게 사고하는 능력을 키우기 위해 제가 더 신경써야 할 부분이 있을까요??

왜냐하면 많은 분들이 딱 여기까지만 하고 끝내구요
(물론 본인만의 생각이 있을 수도 있지만, 아웃풋은 저 문장만 말하거나 말도 안하고 끝남) 

> 제가 생각이 너무 닫혀있었던 건 아닌지 생각했습니다.
코드를 작성하면서 유연한 사고를 하는 건 당연한 것인데 아직 저는 하나를 확실히 하려다가 다른 것을 놓치는 단계에 머물러 있는 것 같습니다. 
꾸준한 연습이 답이라는 걸 저도 알고 있지만. 코드를 작성하면서 유연하게 사고하는 능력을 키우기 위해 제가 더 신경써야 할 부분이 있을까요??

지환님은 이미 근본적인 원인부터 해결방법까지 다 생각을 했으니 해결 방안대로 다른 사람이 말한 내용을 다양하게 시도해보고 직접 경험해보면서 생각의 폭을 넓히면 충분하다고 생각해요. 저는 우테코에서 1단계, 2단계 미션이 지금처럼 1단계 - 자바, 2단계 - 스프링 미션이었는데요. 그때 다른 크루들이랑 제 방법이 다르다면 일단 다른 크루들이 말한 방법으로 뒤갈아 엎었던 기억이 나네요. 

---

제가 회사 초반에 입사기념?으로 임원님이랑 저녁을 먹었을 때, '왜 이 회사에 들어오고 싶었냐', '타회사들보다 사용자가 많은 서비스가 많아 들어오고 싶었다', '아직 걷지도 못하는데 벌써부터 뛰고 싶어 하냐'이런 이야기를 한 적이 있는데요.
저 말을 듣고 '아 다시 바닥부터 시작해야하는구나. 초심을 잃지말고 여기서는 내 레벨이 1이니까 초록달팽이부터 잡아서 경험치를 올리자' 이런 생각으로 살고 있습니다

그런 의미로 [이 글](https://xrabcde.github.io/digging/)을 읽어보는걸 추천할게요
저는 취준할 때 다른 글보다 제일 인상 깊게 읽은 블로그 글인데요. 그래서 그런지 제가 댓글에 쓴 내용이랑 이거랑 비슷하다고 생각합니다

저 글에 나오는 중요하다고 생각하는 문장은 `잡일을 하는 과정을 건너뛰고서는 절대 큰 일을 할 수 없다.`인데, 지환님의 경우에는 잡일과 큰 일을 아래처럼 나눌 수 있을 것 같아요
- 잡일: 다양하게 시도해보기
- 큰 일: 유연한 사고를 하는 일

---

그래서 직접적으로 신경써야 할 부분은 없어보이고, 너무 조급하다는 생각이 든다면 위처럼 조금씩 조금씩 더 길게 보는걸 추천드려요

70825 commented on src/main/java/LottoApplication.java:
```diff
@@ -0,0 +1,9 @@
+
+    public static void main(String[] args) {
+        new LottoController().run();
+    }
```
`new LottoController().run()`으로는 보통 사용하지 않기 때문에 아래처럼 수정하시면 좋아보여요~
```suggestion
    public static void main(String[] args) {
        LottoController lottoController = new LottoController();
        lottoController.run();
    }
```

70825 commented on src/main/java/controller/LottoController.java:
```diff
@@ -0,0 +1,39 @@
+    private static final int PRICE_PER_LOTTO = 1000;
+
+    public void run() {
+        final Price price = InputView.inputPrice();
+        final Lottos lottos = buyLottos(price);
```
`Domain` 패키지에서는 지역변수에 `final`을 사용하지 않던데, `Controller`에서는 `final`을 사용하고 있네요
컨트롤러에서만 `final`을 사용하는 이유가 따로 있을까여??
저는 `final` 사용하는게 좋다고 생각하는데 지환님이 사용하신 의도? 배경? 이런게 궁금하네요

JihwanYeom replied:
Controller 에서는 이미 기능이 구현된 객체들을 이용해 하나의 큰 로직을 만드는 부분이기에 final과 final이 아닌 변수를 잘 구분할 수 있다고 생각하여 사용을 하였는데 이제 보니 domain에서는 그 구분이 잘 되지 않는다고 생각하여 final을 신경 쓰지 않고 변수를 생성한 것 같네요... 

사실 정말 모르겠으면 기능을 완성한 뒤 final을 붙여보기만 해도 불변 여부는 알 수 있을텐데 이 부분은 조금 제가 안일했던 것 같습니다.

70825 replied:
안일하기보다는 보통 코드 줄이 짧은 도메인 로직에서는 지환님 방법이 일반적인걸로 알고 있어요
저 같은 경우엔 `final`을 사용하면 아래의 의도를 가지게 된다고 생각해요.

- final 키워드가 붙여진 곳: 앞으로 이 값이 교체될 일이 없음
(= 나중에 코드 읽을 때 빠르게 읽고 내려갈 수 있음)
- final 키워드가 붙지 않은 곳: 앞으로 이 값은 추후에 교체됨
(= 여기 부분이 더 중요함. 나중에 코드 읽을 때 해당 값을 생각하면서 내려가야하기 때문에)

저는 요즘 코틀린을 사용하고 있는데, 코틀린은 자바와 다르게 `final 키워드 유무`가 아니라 `val`, `var`로 둘 다 명시적인 키워드가 있어서 좋더라구요

70825 commented on src/main/java/controller/LottoController.java:
```diff
@@ -0,0 +1,39 @@
+
+    public Lottos buyLottos(Price price) {
+        List<Lotto> lottoList = new ArrayList<>();
+        for (int i = 0; i < price.getPrice() / PRICE_PER_LOTTO; i++) {
+            lottoList.add(Lotto.of(generator.generateLottoNumbers()));
+        }
```
`Controller`에서 `generator.generateLottoNumbers()`로 랜덤 번호를 생성하고 있는데요
제가 만약 로또 미션을 만든다면 `Lotto`든 `Lottos`든 로또라는 도메인이 랜덤 번호를 생성하도록 만들 것 같은데, 지환님은 컨트롤러에서 랜덤 번호를 생성하게된 의도가 무엇인가요?? (제 이유는 지환님 댓글이 달린 이후에 ㅋㅋ..)

JihwanYeom replied:
아마도 객체를 만들면서 실제로 로또를 하는 상황을 가정해 봤을 때 로또가 스스로 번호를 작성한다기 보다는 외부로부터 번호를 입력받아 번호를 저장하고 있는 역할만 한다고 생각하여 따로 번호를 생성하는 객체를 만든 것 같습니다.
그리고 아마 저번 미션에서 랜덤 변수를 생성했던 방식을 사용했기에 그 방식을 그대로 사용해본 거 같기도 합니다.

사실 저번 미션에서 이런 방식으로 랜덤 변수를 생성한 이유는 테스트가 용이하기 때문이었는데 컨트롤러에서 번호를 생성해서 주입할 경우 굳이 이런 방식으로 생성하지 않아도 테스트에 영향을 크게 주지 않을거라고... 생각하게 되네요! 항상 코드를 작성할 떄 이유를 한번 더 생각해 봐야겠습니다...

70825 replied:
지환님이 생각하는 방향도 맞아요 ㅋㅋ 저건 두가지 관점으로만 딱딱 나뉘어지는 형태라서요

- 컨트롤러가 게임을 진행하는 것이니 로또번호를 생성하는 책임도 컨트롤러에게 있다 (= 현실세계랑 같음)
- 로또가 직접 로또번호를 생성하게 한다

저는 두번째 관점이라 주입을 `generator.generateLottoNumbers()`가 아닌 `generator`라고 생각해요

혹시 `객체지향의 사실과 오해` 이 책 안읽어보셨다면 시간 날 때 읽어보는걸 추천드립니다아

70825 commented on src/main/java/domain/Lotto.java:
```diff
@@ -0,0 +1,30 @@
+import java.util.Collections;
+import java.util.List;
+
+public class Lotto {
```
여기뿐만 아니라 코드가 전반적으로 매우 깔끔하네요 👍👍

70825 commented on src/main/java/domain/Lotto.java:
```diff
@@ -0,0 +1,30 @@
+
+    public static Lotto of(List<LottoNumber> lottoNumbers) {
+        return new Lotto(lottoNumbers);
+    }
```
생성자는 `private`으로 닫아두고, 정적 팩토리 메서드를 사용하니 지환님이 느끼기엔 어떤 장점이 있는 것 같나요??

JihwanYeom replied:
정적 팩토리 메서드를 사용하는 것이 new 를 사용하는 것보다 코드를 보았을 때 더 직관적으로 객체를 생성하는 코드임이 잘 드러나는 것 같았습니다. 그리고 나중에 하나의 객체를 각기 다른 방법으로 초기화 할 때도 있을텐데 그 때 메서드를 확장하면서 메서드마다 역할이 잘 드러나도록 이름을 짓기에 매우 유용할 것이라고 생각합니다!

70825 commented on src/main/java/domain/LottoNumber.java:
```diff
@@ -0,0 +1,33 @@
+        return lottoNumber;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (obj == null || getClass() != obj.getClass())
+            return false;
+
+        LottoNumber lottoNumber = (LottoNumber) obj;
+        return this.lottoNumber == lottoNumber.getLottoNumber();
+    }
```
이 코드는 직접 만드신건가요?? 
이거 물어보는 이유는 저는 만든다면 인텔리제이 eq&hs 코드를 자동 생성해주는 기능을 사용하는데, 그거랑 코드가 살짝 달라가지구요 🧐

JihwanYeom replied:
저도 생성해 주는 기능을 사용하고 equals는 알고 있기에 equals 만 조금 수정해서 사용했습니다... 그리고 hashcode는 필요 없다고 생각되어 삭제 했습니다..만 밑에 리뷰어님 코멘트를 읽고 equals와 hashcode에 대해 공부하고 나니 hashcode 함수도 꼭 필요한 함수였더라고요 그래서 다시 추가했습니다!

70825 commented on src/main/java/domain/Price.java:
```diff
@@ -0,0 +1,23 @@
+        return new Price(price);
+    }
+
+    public Price add(int addPrice) {
+        return Price.from(price + addPrice);
+    }
```
옛날에 저는 이거 생각도 못하다가 다른 사람들 코드 보면서 알게된 내용인데, 이거는 어떻게 생각해서 나오게 된 결과물인가요??

JihwanYeom replied:
이 코드는 제가 생각엔 C++ 의 연산자 오버로딩에서 차용한 느낌이에요 제가 군대에 있을 때 C++을 정말 재밌게 공부했었는데 그 때 가장 매력적이라고 생각했던 도구가 연산자 오버로딩이었어요! 연산자도 하나의 함수로 볼 수 있고 이를 이용해 연산자에 다른 역할을 부여할 수 있다는 색다른 관점이 좋았었다고 해야 할까요..? 그래서 자바는 연산자 오버로딩을허용하지 않지만 충분히 그와 같은 역할을 하는 함수를 만들 수 있다고 생각했어요

70825 replied:
👍👍

70825 commented on src/main/java/view/InputView.java:
```diff
@@ -0,0 +1,32 @@
+
+    private static final Scanner scanner = new Scanner(System.in);
+
+    public static Price inputPrice() {
+        System.out.println("구입금액을 입력해 주세요.");
+
+        return Price.from(scanner.nextInt());
+    }
+
+    public static Lotto inputWinningNumbers() {
+        System.out.println("지난 주 당첨 번호를 입력해 주세요.");
+        scanner.nextLine();
+        List<String> winningNumbers = Arrays.asList(scanner.nextLine().split(","));
+        List<LottoNumber> winningNumberList =
+                winningNumbers.stream()
+                .map(winningNumber -> LottoNumber.from(Integer.parseInt(winningNumber)))
+                .collect(Collectors.toList());
+        return Lotto.of(winningNumberList);
+    }
```
`InputView`가 `Price`, `LottoNumber`, `Lotto`를 생성할 책임을 가지고 있는 것에 대해 어떻게 생각하시나요?!

JihwanYeom replied:
항상 원시값 포장을 유지하고 있어야 한다고 생각해서 InputView 에서도 객체를 반환할 때 포장된 값을 넘겨주었는데
지금 생각해보면 단순히 입력받은 값 만을 가공하여 컬렉션이나 원시값을 그대로 넘겨주고 객체를 생성하는 역할은
컨트롤러에 맡기면 조금 더 편해질 것 같아요 혹시 이 방법이 InputView의 역할에 더 맞는 방법일까요?

70825 replied:
> 단순히 입력받은 값 만을 가공하여 컬렉션이나 원시값을 그대로 넘겨주고
객체를 생성하는 역할은 컨트롤러에 맡기면 조금 더 편해질 것 같아요

저도 그렇게 생각합니다~

70825 commented on src/main/java/view/OutputView.java:
```diff
@@ -0,0 +1,42 @@
+    private static final int MIN_RANK_FOR_PRIZE = 3;
+    private static final int MAX_RANK_FOR_PRIZE = 6;
+
+    public static void printLottos(Lottos lottos) {
```
`InputView`, `OutputView` 메서드들이 전부 `static` 키워드를 사용하더라구요
`static`을 사용하게된 지환님의 생각/배경이 궁금하네여

JihwanYeom replied:
학교에서 과제 프로젝트를 할 때 이곳 저곳에 빈번하게 쓰일 만한 유틸리티 성격이 강한 메서드들을 같은 종류끼리 한 객체에 정리하여 Static으로 사용했던 습관이 그대로 나타난 것 같습니다. 근데 MVC 패턴에선 View에 의존하는 부분은 Controller 밖에 없기 때문에 굳이 Static으로 사용할 필요는 없는 것 같네요... 혹시 더 나아가서 이렇게 Static으로 인스턴스를 생성하지 않고도 기능을 불러올 수 있도록 코드를 작성하는 것이 객체지향적으로 맞지 않은 행동일까요?

70825 replied:
> Static으로 인스턴스를 생성하지 않고도 기능을 불러올 수 있도록 코드를 작성하는 것이 객체지향적으로 맞지 않은 행동일까요?

정말 제대로된 객체지향이라면 맞지 않는 행동이라고 저도 어디선가 책인지 영상인지 본 것 같긴한데, 현실에서 이론이 딱 들어맞는 상황은 거의 없는거라 저는 오히려 어딘가에는 있는게 정상적이라고 생각해요

대신 유틸리티 함수를 어디까지 보냐의 기준이 사람마다 다르지 않을까 생각합니다

이유를 가지고 있는거라면 굳이 코드 수정할 필요는 없을 것 같아요~

70825 commented on src/main/java/domain/Lotto.java:
```diff
@@ -0,0 +1,30 @@
+                .filter(lottoNumber -> winningNumbers.getLottoNumbers().contains(lottoNumber))
+                .count();
+        return LottoRank.from(rankNumber);
+    }
```
> 예를 들면 모든 값을 포장해두면 getter를 자연스럽게 쓰게 되고 이로 인해 해당 객체를 활용하는 코드가 많이 길어진다는
느낌을 받았습니다. 제가 객체의 역할을 잘못 부여해서 생기는 일일까요?

여기서 개선할만한 부분이 보이네요~
`winningNumbers.getLottoNumbers().contains(lottoNumber)`를 쓰기보다는 `contains(...)` 함수를 만들어서 `Lotto` 객체가 더 주도적으로 책임을 질 수 있게끔 만들면 더 좋아보여요

이러면
AS-IS: `winningNumbers.getLottoNumbers().contains(lottoNumber)`
TO-BE: `winningNumbers.contains(lottoNumber)`
이렇게 되서 조금 더 깔끔해지겠네요!

70825 commented on src/main/java/domain/LottoRanks.java:
```diff
@@ -0,0 +1,47 @@
+        return (int) lottoRanks.stream()
+                .filter(lottoRank -> lottoRank.getLottoRank() == rank.getLottoRank())
+                .count();
+    }
```
> 예를 들면 모든 값을 포장해두면 getter를 자연스럽게 쓰게 되고 이로 인해 해당 객체를 활용하는 코드가 많이 길어진다는
느낌을 받았습니다. 제가 객체의 역할을 잘못 부여해서 생기는 일일까요?

`LottoRank` 끼리 비교를 한다면 `equals & hashcode`를 사용하는 편인데요.
동일성과 동등성에 대해 학습해보면 좋아보여요~ (참고로 이건 지환님이 취준할 때 다시 만나게 될 개념입니다)
학습하시고 적용한다면 아래처럼 바뀌는걸 기대해볼 수 있어요

AS-IS : `lottoRank.getLottoRank() == rank.getLottoRank()`
TO-BE : `lottoRank.equals(rank)`
TO-BE2: `return Collections.frequency(lottoRanks, rank)`

70825 commented on src/main/java/view/OutputView.java:
```diff
@@ -0,0 +1,42 @@
+        System.out.println(lottos.count() +"개를 구매했습니다.");
+
+        for(Lotto lotto : lottos.getLottos()) {
+            List<Integer> LottoNumberList =
+                    lotto.getLottoNumbers().stream()
+                    .map(LottoNumber::getLottoNumber)
+                    .collect(Collectors.toList());
```
이건 `Lotto`에서 `List<Integer>`로 반환해주는 getter를 만드는게 낫지 않을까라는 생각이 듭니다
왜냐하면 저는 '`OutputView`가 `Lotto`를 `List<Integer>`로 변환하는 책임을 가지는게 맞나?'라고 생각해봤는데 책임소재가 애매하더라구요
그래서 이지선다로 관점을 바꿔 '`Lotto`를 `List<Integer>`로 바꾸는 로직을 `OutputView`에 두는게 맞을까 아니면 `Lotto`가 반환해주는게 맞을까'를 생각해보면 `Lotto`에 더 가까운 것 같아요
어떻게 생각하시나요?

JihwanYeom replied:
저는 OutputView가 객체를 가져왔으면 이를 어떻게 출력할지는 View 쪽에서 해결해야 한다고 생각했는데. 그 변환의 역할을 Lotto에 넘겨줄 수도 있겠군요! 확실히 그렇게 한다면 코드가 좀 더 깔끔해 질 수 있을 것 같아요. 그런데 저는 getter는 인스턴스 변수를 그대로 넘겨주어야 한다고 생각하고 있었는데. 이런 경우엔 LottoNumber 리스트를 넘겨주는 getter와 Int리스트를 넘겨주는 getter를 따로 만드는 것이 좋을까요...?

70825 replied:
넵넵 저는 따로 만들어두는게 좋다고 생각해요
이건 사람마다 생각이 다를 것 같은데, 저는 `객체 = 사람으로 비유하면 뭐든지 주도적이고 능동적으로 행동하는 사람`이라고 생각해요
객체가 주도적이고 능동적이라면 `List<Integer>`까지 제공할 수 있다고 생각이 들어서요
그래서 이런식으로 생각해서 코드를 작성하게 된다면 아래와 같은 효과를 기대할 수 있습니다
- Controller, Service, View에서는 Domain에 있는 로직들을 조립하는 코드들 위주로 작성하게 댐

70825 reviewed (changes requested):
안녕하세요 지환님~ 코드 레전드 깔끔이네요 👍

> 부족한 부분이 많아 지적을 받을수록 많이 배워가고 있는 것 같습니다. 조금이라도 잘못된 것 같은 코드가 있으면 얘기해 주시면 감사하겠습니다!

잘못된 부분이라고 할만한건 `LottoApplication.java`에서 `new LottoController.run()` 말고는 없는 것 같은데요??!

> 예를 들면 모든 값을 포장해두면 getter를 자연스럽게 쓰게 되고 이로 인해 해당 객체를 활용하는 코드가 많이 길어진다는 느낌을 받았습니다. 제가 객체의 역할을 잘못 부여해서 생기는 일일까요?

이거는 조금이라도 지환님이 getter를 덜 쓸 수 있게 개선할만한 부분 코멘트 달아두었어요~!!

> 그리고 포장된 객체들의 리스트를 다루는 것도 고민이 많이 되었습니다 포장된 객체들의 리스트를 가공하거나
출력하기 위해 stream을 많이 사용했는데 이것이 정말 좋은 코드인지, 너무 복잡하게 짜는 코드는 아닌지
고민했던 것 같습니다.

제가 생각하기엔... 이게 진짜 미션의 목적 아닐까요??

저는 우테코 미션할 때 내적 갈등이 많이 일어난 부분이 `이렇게까지 해야 해???`였는데요 🤣

근데 나중에 팀프로젝트하면서, 회사 다니면서, 지금 코드리뷰하면서 다시 nextstep 사이트 들어가보니 느낀점은 `어떻게든 투머치하게 해봐야 나만의 제대로된 생각을 가질 수 있게 되구나, 나중에 코딩할 때는 어느정도 적당한 선을 판단할 수 있게 되구나`라는 생각이 들더라구요

저는 그래서 지환님이 이런 고민을 한다는 자체가 굉장히 좋은 방향으로 학습하고 있다는걸 증명하고 있다고 생각합니다~

1,2 단계 미션에서는 투머치하게 해도 나중에 3,4단계에서는 별로라고 생각하는 부분을 수정해보는 것도 좋아보여요. 이러면 둘 다 경험하게 되니까 지환님의 확실한 근거를 챙길 수 있는 장점이 있어보입니다

> 그리고 Rank에 대한 가격을 상수로 선언해 두었는데 이렇게 어떤 수에 따른 결과값을 상수로 선언하는 것이 좋은 방법은 아닌 것 같습니다.

이거는 3단계에서 `Java Enum을 적용한다`라는 요구사항이 있으니 해결될 것으로 보입니다..??
아니라면 3,4단계 PR에서 다시 이야기 해보시죠..!!

---

## 코멘트 외의 개선사항

### 1. 테스트 코드 (1, 2단계에 무조건 추가)

디스코드 > 백엔드-자유 채널 > 루카가 올린 내용 참조

저 내용에 추가로 이야기하자면 제가 학교 다니면서 느꼈던건 지금 이런 활동 아니면 테스트 코드에 대한 경험을 얻어가기 굉장히 어려워요 🫠 

특히 저는 우테코 끝나고 학교에 복학하니 테스트 코드를 작성해본 사람과 안해본 사람의 차이를 매우 크게 느꼈기도 하구요

그래서 지환님도 빠르게 테스트 코드 작성에 숙달이 되서 테스트 코드에 대한 인사이트도 많이 넓히면 좋겠다는 생각이 듭니다. 특히 테스트 코드는 나중에 TDD, ATDD를 자연스럽게 적용할 정도의 수준이 된다면`지금보다 더 좋은 코드/구조를 어떻게 만들 수 있을까`쪽으로 고민하는데 많은 도움이 되어 있을 거에요

저는 프로페셔널한 개발자가 되려면 어떻게 해야할까 생각해보자면 제일 중요한건 남의 시간을 소중히 생각하고, 책임감 있는 사람이라고 생각하는데요. 그 시작이 테스트 코드라고 생각합니다. 왜냐하면 대학교에서의 일반적인 FE-BE 프로젝트는 QA 짬처리를 보통 FE가 하게되는 불상사가.... ㅋㅋㅋ

아무튼 저는 가독성 좋은 코드도 중요하지만, 테스트 코드랑 저울질 해보면 테스트 코드가 더 우선순위가 높다고 생각해요

그래서 열심히 만드시길 바랍니다 💪💪


### 2. 미션 추가 요구사항 (3, 4단계에서 진행해도 무방)

[1] 현재 입력값은 무조건 `숫자,숫자,숫자,숫자,숫자,숫자`로 붙여서 입력을 해야 오류가 발생하지 않더라구요

그래서 아래의 입력값들이 모두 정상 동작할 수 있도록 제약조건을 추가해 봅시다~

- `1,2,3,4,5,6 (현재 통과중)`
- `1, 2, 3, 4, 5, 6 (에러 발생)`
- `1,   2,       3,4,   5,   6 (에러 발생)`

두번째 데이터가 통과하면 세번째 데이터는 자동으로 통과할거라고 기대해요

[2] 로또 숫자가 6개가 딱 들어오지 않을 경우 여기에 대한 에러 처리가 필요해보입니다.
(ex. 5개가 들어온다거나, 7개가 들어온다거나..)

재입력을 할지, 에러 발생하고 프로그램을 종료할지는 지환님이 시간이 되는데로 적용해주시길 바래요~

JihwanYeom committed:
refactor : main 메서드 및 공백 입력 시 오류 처리

main 메서드에서 Controller 객체 생성과 메서드 호출 분리

로또 당첨 숫자 입력 시 공백이 있어도 오류 없이 입력되도록 수정

JihwanYeom committed:
refactor : InputView 에 부여된 객체 생성 책임을 객체에게 부여

JihwanYeom committed:
refactor : getter 사용을 줄이기 위한 수정

Lotto 클래스에서 Conatins 메서드를 정의하여
getter 사용을 한단계 줄임

LottoRank 클래스에서 hashCode 메서드까지 정의하여 랭크를 구하는 로직을 간소화

JihwanYeom committed:
refactor : Lotto 에 정수 리스트를 반환하는 getter 구현

JihwanYeom committed:
refactor : OutputView 에서 로또 숫자를 가공하는 책임 제거

JihwanYeom committed:
chore : Price 에 divideBy 메서드를 만들어 더 간결하게 코드 표현

JihwanYeom committed:
test : 로또 랭크 개수 구하기 테스트 구현

JihwanYeom committed:
chore : 푸시 전 점검

JihwanYeom commented:
안녕하세요 리뷰어님! 늦은 시간에 커밋 올리게 되어 죄송합니다 리뷰해주신 내용 적용해 보았습니다!
그리고 제가 미처 보지 못한 리뷰들도 있었더라고요... 리뷰어님께서 너무 정성스럽게 리뷰를 써주셔서 제가 접혀진 리뷰들을 미처 발견하지 못했습니다... 그래서 답변하지 않았던 리뷰들에도 답변 달았습니다! 생각해볼 만 한 질문 던져주셔서 큰 도움이 되었어요!

먼저 리뷰어님께서 구현해보라고 하셨던 입력시 공백 처리는 trim 메서드를 이용해 해결해 보았습니다!

그리고 InputView 에 부여되어있던 책임을 객체에게 부여했는데 이 과정에서 InputView가 굉장히 깔끔해지고 객체의 역할도
훨씬 더 확실해진 것 같습니다

또한 getter 사용을 줄이기 위해 equal 이나 contains 같은 함수를 정의하는 것도 굉장히 도움이 많이 되었습니다!
처음에는 이렇게 자바가 제공하는 객체의 기본적인 성질을 바꾸는 걸 굉장히 조심스럽게 생각하였는데
잘 사용하면 코드가 훨씬 직관적으로 변한다는 것을 알게 된 것 같습니다!

그리고 메서드를 통해 객체를 주고 받을 때는 굳이 꼭 포장된 값을 사용하지 않아도 된다고 생각하고 다시 코드를
작성해보니 훨씬 로직이 잘 보이게 되었습니다.
사실 지금 생각해보면 지금 제 LottoRanks 가 하는 역할이 랭크를 관리하고 있는 역할이라는 걸 또 깨닫게 되었습니다..
그래서 굳이 다른 객체를 추가하지 않고도 이해할만한 코드가 되었다고 생각합니다! 아니면 다음 단계에서는
객체의 이름을 바꿔보는 시도도 해볼까 합니다.

사실 제일 고민을 많이 한 건 테스트 코드 부분입니다. 특별한 역할이 없는 정적 팩토리 메서드나, getter 메서드 등을 제외하고
핵심적인 로직이라고 생각하는 메서드들을 테스트하였는데 너무 간추려서 한 것은 아닌가 걱정이 됩니다.

그리고 테스트를 하는 방식이 올바른 방법인지도 조금 확신이 잘 안 섭니다... AssertJ 를 이용한 테스트 코드 작성을
저번 미션에 배웠는데 제가 AssertJ 를 아직 전부 활용하지 못하고 있다는 생각이 듭니다... 리뷰어님의 테스트 코드 작성 단위
기준과, 의미 있는 테스트 코드를 작성하기 위해 어떤 점을 지켜야 하는지 조금 힌트를 주신다면 감사하겠습니다!

70825 commented on src/main/java/controller/LottoController.java:
```diff
@@ -0,0 +1,39 @@
+
+    public Lottos buyLottos(Price price) {
+        List<Lotto> lottoList = new ArrayList<>();
+        for (int i = 0; i < price.divideBy(PRICE_PER_LOTTO); i++) {
```
👍

70825 commented on src/test/java/ControllerTest.java:
```diff
@@ -0,0 +1,21 @@
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+public class ControllerTest {
```
이거는 `test/java`에 전부 테스트 파일들이 있는데요. `main/java`에 있는 패키지 구조로 만들면 더 좋습니다

맥북 기준으로 `LottoController.java`에서 `Cmd + Shift + T`를 누르고, `Create New Test` > `OK`를 클릭하면 인텔리제이에서 자동으로 테스트 파일을 생성해줍니다. 윈도우 쓰시면 거기에 맞는 단축키가 있을거에요

70825 commented on src/test/java/ControllerTest.java:
```diff
@@ -0,0 +1,21 @@
+
+public class ControllerTest {
+
+    @DisplayName("로또 구매 테스트")
+    @Test
```
여기는 아래 내용 한번 고민해보시면 좋아보여요
`어노테이션 순서에도 컨벤션을 적용할 수 있을까? 있다면 어떤 순서대로 하는게 좋을까?`

70825 replied:
그리고 `@DisplayName`에서는 `로또 구매 테스트`보다는 `로또를 3000원치 구매하면 3개를 구매할 수 있어야한다`라고 원인과 결과를 적어두시면 더 좋습니다~

JihwanYeom replied:
어노테이션 순서에 대해 고민을 해봤는데... 제 생각에는 더 메서드에 많이 관여하는 핵심적인 어노테이션 일수록 앞에 와야 한다고 생각하게 된 것 같아요 그래서 영향성을 생각해 봤을 때 Test가 DisplayName 보다 위에 오도록 변경하였습니다!

70825 commented on src/test/java/ControllerTest.java:
```diff
@@ -0,0 +1,21 @@
+
+    @DisplayName("로또 구매 테스트")
+    @Test
+    public void testBuyLottos() {
+        LottoController testController = new LottoController();
+        Price testPrice = Price.from(3000);
+        Lottos testLottos = testController.buyLottos(testPrice);
+
+        assertThat(testLottos.getLottos().size()).isEqualTo(3);
+    }
```
여기는 `given - when - then` 검색해서 적용해보면 좋아보여요~
```suggestion
    public void testBuyLottos() {
        // given
        LottoController testController = new LottoController();
        
        // when
        Price testPrice = Price.from(3000);
        Lottos testLottos = testController.buyLottos(testPrice);

        // then
        assertThat(testLottos.getLottos().size()).isEqualTo(3);
    }
```

그리고 테스트 코드는 결국엔 then 부분이 중요하니까 아래처럼 `actual`, `expected` 값까지 추가하면 다른 사람이 코드를 볼 경우 더 쉽게 파악이 가능해요 (나중에 역방향으로 추적하기 쉬움)

```java
    public void testBuyLottos() {
        // given
        LottoController testController = new LottoController();
        int expected = 3;
        
        // when
        Price testPrice = Price.from(3000);
        Lottos testLottos = testController.buyLottos(testPrice);
        int actual = testLottos.getLottos().size();

        // then
        assertThat(actual.isEqualTo(expected);
    }
```

JihwanYeom replied:
테스트 코드를 작성할 떄 이 패턴을 작성 하면 되는군요! 확실히 변수로 actual 값과 expected 값을 명시해 주니 훨씬 깔끔하고 알아보기 쉽네요! 여기서는 하나의 값을 테스트 하는데 사용되었는데 많은 값을 테스트 해야 할 경우에도 적용할 방법이 있을까요?

70825 replied:
저 같은 경우에는 그냥 뒤에 숫자 붙이고 있어요 ㅋㅋ
`expected1`, `expected2` 이것처럼요

70825 commented on src/test/java/LottoNumberTest.java:
```diff
@@ -0,0 +1,17 @@
+    public void testEquals() {
+        LottoNumber testNumber = LottoNumber.from(1);
+
+        assertThat(testNumber.equals(LottoNumber.from(1))).isEqualTo(true);
+        assertThat(testNumber.equals(LottoNumber.from(2))).isEqualTo(false);
```
1) 여기 두 개는 각각 `true`, `false` 기대값이 다르므로 메서드를 분리해서 테스트하는게 좋아요
2) `isEqualTo(true)` 대신 `isTrue()`, `isFalse()`로도 축약 가능해요~

70825 commented on src/test/java/LottoRanksTest.java:
```diff
@@ -0,0 +1,42 @@
+
+        LottoRanks testLottoRanks = LottoRanks.of(rankList);
+
+        assertThat(testLottoRanks.getNumberOfRank(LottoRank.from(3))).isEqualTo(2);
+        assertThat(testLottoRanks.getNumberOfRank(LottoRank.from(5))).isEqualTo(1);
+        assertThat(testLottoRanks.getNumberOfRank(LottoRank.from(6))).isEqualTo(1);
```
`assertThat` 여러개를 검증한다면 `softAssertions`를 사용하는걸 추천해요
`assertThat`은 첫줄이 틀리면 두세번째 테스트는 테스트를 안하는데, `softAssertions`는 일단 모든 테스트를 검증해서 어디어디가 틀렸는지 나오게 됩니다~

## 비교
값에 전부 +1을 해서 모든 테스트가 틀려야하는 상황
![스크린샷 2025-03-30 오전 10 31 53](https://github.com/user-attachments/assets/fcb70316-3196-4074-a9bf-18ebfd756d50)

### Assertions
첫번째만 틀렸다고 나옴

![스크린샷 2025-03-30 오전 10 32 29](https://github.com/user-attachments/assets/54645166-762a-4035-a435-0a5feb48f1cc)

### SoftAssertions

```java
        assertSoftly(soft -> {
            soft.assertThat(testLottoRanks.getNumberOfRank(LottoRank.from(3))).isEqualTo(3);
            soft.assertThat(testLottoRanks.getNumberOfRank(LottoRank.from(5))).isEqualTo(2);
            soft.assertThat(testLottoRanks.getNumberOfRank(LottoRank.from(6))).isEqualTo(2);
        });
```

3개가 틀렸다고 나옴

![스크린샷 2025-03-30 오전 10 33 18](https://github.com/user-attachments/assets/5e6deeeb-7043-43c3-8138-6a6393848eab)

70825 commented on src/main/java/controller/LottoController.java:
```diff
@@ -0,0 +1,39 @@
+        OutputView.printRateOfReturn(totalPrize, price);
+    }
+
+    public Lottos buyLottos(Price price) {
```
지금 보니 여기가 `public`으로 열려있었네요 😅
지금은 테스트 코드 때문에 `public`으로 열 수 밖에 없긴한데, 프로덕션 코드는 `private`로 사용해도 무방하니 지금 상황이면 `private`로 사용하는게 맞을 것 같아요

그러면 
1) `private`로 바꾸기 > 테스트가 매우 어려워짐
2) `public`을 유지하는 방법 > 테스트가 용이하지만 어딘가로 이동이 필요함
둘 중에 하나는 만족해야하는데, `buyLottos()`를 `Lottos.java`로 이동하는 2번 방법이 좋아보입니다. 이러면 `public`으로 사용할 명분이 있어보여요

70825 reviewed (changes requested):
지환님 고생하셨습니다~ 🥳🥳

저는 이제 Approve를 줘도 매우매우 충분하다고 생각이 드는데요. 내일 3,4단계 스터디 시작이라 저는 내일 아침에 무조건 Approve할 예정이니 지환님 선택대로 내일 아침까지 수정해주시거나, 3,4단계에서 반영해주시면 됩니다~

### 현재 고쳐야할 사항
- 아래 코멘트 내용들

### 3,4 단계 제출시 추가 요구사항
지금처럼 `Domain`에 있는 아래 조건을 제외한 모든 public 메서드는 테스트 코드를 작성해주세요~
  - 정적 팩토리 메서드 제외
  - 단순 getter 제외

메서드들이 잘 테스트 되고 있는지?에 대한 내용은 아래처럼 특정 패키지나 파일에 마우스 오른쪽 버튼 클릭해서  IDE 기능 사용하시면 돼요

![스크린샷 2025-03-30 오전 10 59 24](https://github.com/user-attachments/assets/05192f04-0ae0-4247-aca0-5e9cab14f1a4)

- Run 'ㅇㅇㅇ' with Coverage 클릭
- 이후 다시 `Run 'ㅇㅇㅇ' with Coverage 클릭`를 하면 버튼 3개가 있는 창이 추가로 나오는데 `Replace Active Suites` 클릭

이러고 프로덕션 코드들 보시면 초록색이랑 빨간색으로 어디까지 테스트 되고 있는지 확인할 수 있어요~!!

> 테스트를 하는 방식이 올바른 방법인지도 조금 확신이 잘 안 섭니다... AssertJ 를 이용한 테스트 코드 작성을
저번 미션에 배웠는데 제가 AssertJ 를 아직 전부 활용하지 못하고 있다는 생각이 듭니다... 리뷰어님의 테스트 코드 작성 단위
기준과, 의미 있는 테스트 코드를 작성하기 위해 어떤 점을 지켜야 하는지 조금 힌트를 주신다면 감사하겠습니다!

이거는 지환님이 오늘 고친다면 고칠 수 있는 중요한 부분들 위주로 코멘트 했어요
나머지는 3, 4단게에서 더 이야기 하겠습니다 🚀🚀

JihwanYeom committed:
refactor : buyLottos 메서드 이동

buyLottos 메서드를 LottoController 에서
삭제하고 Lottos 객체의 정적 팩토리 메서드 형태인 Lottos.buy() 메서드로 변경함

JihwanYeom committed:
test : 테스트 코드 관련 리뷰 반영

테스트 코드 패키지 구조 변경

단일 값 비교시 expected, actual 변수를 이용해 가독성 향상

여러 값 비교시 assertSoftly 를 이용해 테스트가 실패한 부분을 추적 가능하도록 변경

기댓값이 다른 테스트 분리

@Test가 @DisplayName보다 앞에 오도록 변경

JihwanYeom commented:
테스트 코드에 대한 리뷰까지 반영 완료했습니다! 모든 리뷰 하나하나가 정말 큰 공부가 되는 것 같습니다.
늘 자세한 리뷰 감사합니다!! 다음 단계도 잘 부탁드립니다!

70825 reviewed (approved):
고생하셨습니다 🥳🥳🥳
3, 4단계도 화이팅~!!
